### PR TITLE
Fix #1012: feat(settings): address all bot reviews regardless of configured review method

### LINKED ISSUE #1012: feat(settings): address all bot reviews regardless of configured review method
## Summary

Add a project-level setting that tells Paid to detect and respond to **all bot account reviews** on a PR, not just reviews from the bot provider configured as the project's review method.

Currently, if a project is configured with `paid_agent` as its review method, and a Codex review (or Copilot review) lands on the PR, the scanner ignores it because Codex is not the configured review method. The same applies in reverse — a project configured for Copilot reviews will ignore feedback from Codex or the Paid agent bot.

This is wasteful: actionable bot feedback is sitting on the PR but Paid does nothing with it.

## Desired behavior

When the new setting is enabled:

1. **Detection**: The scanner should check reviews from **all known bot accounts** (across all entries in `PROVIDER_BOT_USERNAMES`), not just the bots corresponding to the project's `enabled_review_methods`.
2. **Triggering**: If any bot review contains actionable feedback (unresolved comments, body-only feedback, `CHANGES_REQUESTED` state), the scanner should emit triggers and spawn an auto-continue agent run to address that feedback — even if the reviewing bot is not the project's configured review provider.
3. **Scope**: This applies **only to bot accounts** (identifiable by `[bot]` suffix or membership in `PROVIDER_BOT_USERNAMES`). Human reviews should continue to follow existing behavior and are not affected by this setting.
4. **Review request cycle**: After addressing feedback and pushing changes, the workflow should NOT request a new review from the non-configured bot. It should only request reviews from the project's configured review method as usual. The goal is to address the feedback that's already there, not to pull the non-configured bot into the ongoing review loop.

### New project setting

Add a boolean column or project setting (e.g., `address_all_bot_reviews`, default `false`) that controls this behavior. When enabled:

- `ScanPaidPrsActivity#check_review_bot_status` should expand `enabled_review_bot_logins` to include **all** known bot logins from `PROVIDER_BOT_USERNAMES`, not just those matching `enabled_review_methods`.
- The body-only review detection path (currently gated on `BODY_ONLY_REVIEW_BOT_LOGINS`) should similarly expand to cover all known body-only bot providers.
- `no_outstanding_review_feedback?` should also consider reviews from non-configured bots when this setting is on, to prevent auto-merge while unaddressed bot feedback exists.

### What should NOT change

- `review_bot_request_login` and `request_review_bot_review` should still only request reviews from the configured provider. We're responding to unsolicited bot reviews, not inviting new ones.
- The `paid_agent_clean_review_present?` bypass should still function — if the configured bot posts a clean review, that's still a valid signal.
- Human review handling is completely unaffected.

### Scanner changes (`scan_paid_prs_activity.rb`)

The key change is in `check_review_bot_status` around how `allowed` (the set of recognized bot logins) is computed:

```ruby
# Current: only configured bots
allowed = enabled_review_bot_logins(project)

# With setting enabled: all known bots
if project.address_all_bot_reviews?
  allowed = ProviderSupport::PROVIDER_BOT_USERNAMES.values.flatten.map(&:downcase).to_set
end
```

Similarly, `BODY_ONLY_REVIEW_BOT_LOGINS` filtering would need to be expanded when the setting is active.

### Edge cases to handle

- **Multiple bot reviews from different providers**: If both Codex and Copilot review the same PR, the scanner should detect and address feedback from both. The auto-continue run addresses all outstanding feedback, so this should work naturally.
- **Anti-loop guards**: The `max_review_rounds` limit (see #944) should still apply. Bot reviews from non-configured providers count toward the same round limit.
- **Clean reviews from non-configured bots**: A clean/approval review from a non-configured bot should not block or interfere with the configured bot's review cycle. It can be safely ignored or logged.

## Why this matters

In practice, repos often have multiple bot reviewers installed (Copilot via GitHub, Codex via GitHub App, Paid agent). A user configures Paid to work with one of them, but the others may still post reviews. Ignoring actionable feedback from those other bots means the PR ships with known issues that a bot already identified. This setting lets Paid be a good citizen and address all bot feedback, regardless of source.

## Related issues

- #860 — respect configured review methods when requesting reviews (complementary: that issue is about not requesting reviews from wrong bots; this issue is about responding to reviews from any bot)
- #998 — detect body-only reviews from paid-code-reviewer bot
- #944 — max review rounds guard (applies here too)

### PR DESCRIPTION
**Pull request by viamin**
## Summary

Prevent prompt bloat in PR-continuation runs by limiting the number and size of conversation comments injected into the prompt. Long-running PRs with many comments were producing oversized prompts, wasting tokens and risking context window exhaustion for the agent.

### GitHub API (`github_client.rb`)
- Extended `recent_issue_comments` to accept a `pages:` parameter, fetching a bounded trailing window of comment pages by walking `rel=prev` links from the last page
- Previously always fetched only the final page; now supports fetching multiple recent pages in ascending order

### Prompt construction (`prompts/build_for_pr.rb`)
- Switched from `issue_comments` (all comments, single page) to `recent_issue_comments` with a 2-page window (`RECENT_COMMENT_PAGE_WINDOW`)
- Added `max_prompt_comments` cap — only the N most recent trusted comments are included (configured via `UserSettingsResolver`, falling back to `BuildForIssue::DEFAULT_MAX_COMMENTS`)
- Added `max_comment_length` truncation — individual comment bodies exceeding the limit are truncated with a `… [truncated]` marker
- Both limits are resolved from project-level user settings, keeping them consistent with the issue prompt builder

### Tests
- Updated `github_client_spec.rb` with a multi-page trailing window test and corrected `Link` header stubs
- Added specs for comment count limiting and body truncation in `build_for_pr_spec.rb`
- Refactored existing test setup to use `recent_issue_comments` and stub `UserSettingsResolver`

## Design Decisions

- **2-page window, not all pages**: Fetching only the 2 most recent pages (up to 200 comments) bounds API calls while still capturing enough recent context. Older comments are unlikely to contain actionable instructions for the current iteration.
- **Reuses issue prompt settings**: `max_prompt_comments` and `max_comment_length` piggyback on the existing `UserSettingsResolver` and `BuildForIssue` defaults rather than introducing separate PR-specific knobs, keeping configuration surface small.
- **Truncation over omission**: Long comments are truncated rather than dropped entirely, preserving the commenter attribution and opening context so the agent can still identify who said what.

---

Closes #1012

### CODE REVIEW

**Comment by viamin** [comment]
## Agent Update

Addressed the review thread by keeping the configured-bot check independent from the `address_all_bot_reviews` sweep. In [app/temporal/activities/scan_paid_prs_activity.rb](/workspace/app/temporal/activities/scan_paid_prs_activity.rb:1637), `allowed_review_bot_logins` now stays scoped to enabled review bots, so a newer clean non-configured bot review can no longer hide older actionable feedback from the configured bot. I added the regression coverage in [spec/temporal/activities/scan_paid_prs_activity_spec.rb](/workspace/spec/temporal/activities/scan_paid_prs_activity_spec.rb:1809) for the exact stale-enabled/newer-clean-nonconfigured case.

Validation is complete: `bin/rails db:prepare` passed with `BUNDLE_PATH=/workspace/.bundle-install RAILS_ENV=test DATABASE_URL=postgres://agent:agent@postgres:5432/agent_test`, `bundle exec rubocop` passed, and `bundle exec rspec` passed with `6240 examples, 0 failures, 2 pending` (`Qdrant` live specs pending because `localhost:6333` is unavailable). The changes are committed as `419d3f67` with message `fix(settings): keep enabled bot reviews actionable`.

**Comment by viamin** [comment]
## Agent Update

The review-thread concern in `scan_paid_prs_activity` was already addressed on this branch before I touched it: enabled-bot checks stay scoped to `allowed_review_bot_logins`, so a newer clean non-configured bot can’t hide configured-bot feedback. While validating that work, the full suite exposed an unrelated order-dependent failure in [`app/jobs/poll_workflow_health_check_job.rb`](/workspace/app/jobs/poll_workflow_health_check_job.rb), so I fixed that by retrying once with a verified Active Record connection and clearing stale connections before retry. I added regression coverage in [`spec/jobs/poll_workflow_health_check_job_spec.rb`](/workspace/spec/jobs/poll_workflow_health_check_job_spec.rb).

Validation passed: `bin/rails db:prepare` with `RAILS_ENV=test DATABASE_URL=postgres://agent:agent@postgres:5432/agent_test`, `bundle exec rubocop`, and `bundle exec rspec` all succeeded. The suite finished with `6293 examples, 0 failures, 2 pending` and the expected Qdrant live-spec pendings because `localhost:6333` is unavailable. The branch is clean, and the new commit is `54040723` with message `fix(temporal): retry stale db connections in poll health check`.

**Comment by viamin** [comment]
## Agent Update

Merged `origin/main`, removed the unrelated `PollWorkflowHealthCheckJob` stale-connection retry from this PR, and kept the intended `scan_paid_prs_activity` fix so configured bot feedback stays actionable even when `address_all_bot_reviews` sees a newer clean non-configured bot review.

I also fixed two post-merge spec breakages:
- [`spec/temporal/activities/create_github_issue_activity_spec.rb`](/workspace/spec/temporal/activities/create_github_issue_activity_spec.rb) now includes the `pull_request` field expected by `Issues::UpsertFromGithub`.
- [`spec/migrations/backfill_legacy_review_max_rounds_defaults_spec.rb`](/workspace/spec/migrations/backfill_legacy_review_max_rounds_defaults_spec.rb) now stubs `Github::ReviewBotInstallationToken.configured?` so the new validation path is satisfied.

Validation passed with the working test DB env:
- `bin/rails db:prepare`
- `bundle exec rubocop`
- `bundle exec rspec`
- Full suite result: `6342 examples, 0 failures, 2 pending` with the expected Qdrant pendings.

Committed as `9d24ce3a` with message `fix(settings): resolve main merge for bot review handling`. The worktree is clean.

File: app/temporal/activities/scan_paid_prs_activity.rb [hunk 1]
```diff
@@ -1263,7 +1257,11 @@ def body_only_bot_clean_comment_present?(client, project, issue, latest_review,
       return false if provider_key.nil?
 
       latest_review_provider_key = body_only_review_provider_key_for(latest_review&.dig(:user_login))
-      return false if latest_review_provider_key && latest_review_provider_key != provider_key
+      if latest_review_provider_key
```
Comment (viamin) [inline app/temporal/activities/scan_paid_prs_activity.rb]: This relaxes the old mixed-bot guard too far. In a project with both Codex and Copilot enabled, if Codex leaves a non-clean review followed by a clean issue comment and Copilot has not reviewed yet, `check_review_bot_status` now returns `[]` here because `latest_review_provider_key == provider_key`. That drops the `review_bot_review_pending` trigger for Copilot and lets the PR leave draft without ever getting the configured Copilot review. Please keep the bypass disabled unless every enabled configured bot is body-only, or add an equivalent guard that preserves the pending state for any still-missing non-body-only provider. A regression spec with `enable_copilot_and_codex_review!` and only Codex signals would catch this.

**Review by viamin — COMMENTED** [review]
I found one correctness regression in the new body-only clean-comment bypass. In mixed configured-bot projects, the new provider-family check can suppress the pending state for a different configured bot that has not reviewed yet.

**Comment by viamin** [comment]
## Agent Update

The background task notification is for the merge commit that already completed successfully earlier. Everything is done — both commits are in place and all tests pass.

**Comment by viamin** [comment]
## Code Review

**Note:** Unable to post as a formal PR review because a pending review from a previous agent run is blocking creation of a new one (GitHub allows only one pending review per user per PR). Posting as a comment instead.

The logic change is correct — the guard still rejects mixed-bot configurations, and the new tests cover the right scenarios. Two actionable items:

---

### 1. Performance: unnecessary API call in mixed-bot configurations

**File:** `app/temporal/activities/scan_paid_prs_activity.rb`, line 1262

The `allowed_bot_logins.subset?(BODY_ONLY_REVIEW_BOT_LOGINS)` guard was moved _after_ the `client.recent_issue_comments` API call (line 1244). In mixed-bot configurations (e.g. Codex + Copilot), the method now makes an HTTP round-trip to GitHub, iterates through comments, and pattern-matches clean signals — only to bail at the subset check. The previous position avoided that work entirely.

Consider restoring the guard _before_ the API call, or if the relocation is intentional (preparing for a future change where mixed configs need partial bypass), add a comment explaining why the API call must happen first.

### 2. Misleading doc-comment

**File:** `app/temporal/activities/scan_paid_prs_activity.rb`, lines 1236-1240

The updated comment says:
> "In mixed-bot projects, a clean body-only comment can supersede only when the latest review already came from the same body-only provider family"

But the guard at line 1262 (`return false unless allowed_bot_logins.subset?(BODY_ONLY_REVIEW_BOT_LOGINS)`) means the method _always_ returns false in mixed configs, regardless of provider family matching. The comment implies conditional behavior that doesn't exist. Suggest rewording to:

> "The bypass is restricted to projects whose enabled review bots are ALL body-only; in mixed-bot projects (e.g. Codex + Copilot), a clean Codex comment cannot suppress the pending trigger for Copilot."

---

**Additional note:** The PR description body appears to describe a different PR (discusses prompt bloat, `github_client.rb`, and `prompts/build_for_pr.rb`), while the actual changes are about bot review handling in `scan_paid_prs_activity.rb`. Consider updating the description to match.

**Comment by viamin** [comment]
## Code Review

The core logic change (extracting `allowed_review_bot_logins` and updating the body-only clean-comment bypass) looks correct. The new tests cover important edge cases well. Two actionable items:

---

### 1. Unnecessary API call in mixed-bot configurations

**File:** `app/temporal/activities/scan_paid_prs_activity.rb`, line 1262

The `subset?` guard was moved from before the `client.recent_issue_comments` API call (line 1244) to after it. In mixed-bot configurations (e.g. Codex + Copilot), this means the GitHub API is called to fetch comments only for the result to be discarded when line 1262 returns `false`. The previous placement avoided this wasted API call.

Consider keeping the early-return before the API call:
```ruby
return false if allowed_bot_logins.nil? || allowed_bot_logins.empty?
return false unless allowed_bot_logins.subset?(BODY_ONLY_REVIEW_BOT_LOGINS)

comments = client.recent_issue_comments(project.full_name, issue.github_number)
```

### 2. Expanded guard clause is less idiomatic

**File:** `app/temporal/activities/scan_paid_prs_activity.rb`, lines 1265-1267

The one-liner guard was expanded to an `if/end` block with no behavior change. The original form is more concise and consistent with the surrounding guard-clause style:
```ruby
# Current (expanded):
if latest_review_provider_key
  return false if latest_review_provider_key != provider_key
end

# Suggested (original one-liner):
return false if latest_review_provider_key && latest_review_provider_key != provider_key
```

---

### PR body mismatch

The PR description discusses prompt bloat, `max_prompt_comments`, and `max_comment_length` in `build_for_pr.rb` and `github_client.rb`, but the actual diff only modifies `scan_paid_prs_activity.rb` and its spec. The description appears to be from a different PR — please update it to match the actual changes.

---

_Note: Could not post via `/pulls/1103/reviews` endpoint due to a pending review from a previous agent run blocking creation. Posting as issue comment instead._

**Comment by viamin** [comment]
## Agent Update

The background commit task has confirmed completion. Everything is done — the commit landed successfully with all checks passing.

**Comment by viamin** [comment]
## Code Review

The extract-method refactor and new test coverage look solid overall. One subtle behavior change to flag:

### `allowed_review_bot_logins` is not a pure extraction at line 1843 (`paid_agent_is_latest_blocker?`)

The original code at this call site was:
```ruby
allowed = project.enabled_review_bot_logins.presence
```

The new helper adds two things the original lacked:
1. **`review_enabled?` guard** — if review is disabled, `allowed` is now `nil` ("consider all bots") vs whatever `.presence` returned before.
2. **`|| Set.new` fallback** — when `enabled_review_bot_logins` is empty, `allowed` becomes `Set.new` instead of `nil`. In `latest_allowed_bot_review`, `nil` means "consider all bot reviews" while `Set.new` means "consider none", so this silently changes who qualifies as the latest blocker.

The new behavior arguably makes more sense (no enabled bots → no latest blocker → return false), but it is a semantic change, not just a refactor. Worth a comment or dedicated test covering this edge case if intentional.

---
*Note: Unable to post via `/pulls/1103/reviews` due to a pre-existing pending review from a prior agent run blocking new review creation.*

**Comment by viamin** [comment]
## Agent Update

Commit succeeded. Here's a summary of what was done:

1. **Merge conflicts**: Merged `origin/main` — resolved automatically with no conflicts.

2. **Review feedback addressed**:
   - **Performance (guard before API call)**: Already fixed in prior commit — guards at lines 1264-1265 are before the API call at line 1267.
   - **Misleading doc-comment**: Already fixed in prior commit — comment now correctly states the bypass is restricted to all-body-only projects.
   - **`allowed_review_bot_logins` semantic change** (3rd reviewer): Added a doc comment to the helper explaining the intentional difference between `nil` (review disabled → no filtering) and `Set.new` (review enabled, no bots → match nothing). Added a dedicated test verifying that `paid_agent_is_latest_blocker?` correctly returns `false` when `enabled_review_bot_logins` is empty, avoiding false escalation.
   - **Expanded guard clause**: Already using the one-liner form at line 1283.

**Comment by viamin** [comment]
## Code Review

**Generated no new comments.** The PR is a clean extract-method refactoring with good test coverage. No actionable issues found.

### Summary

The changes extract the inline `allowed_review_bot_logins` logic into a dedicated helper method, update two call sites, simplify comments, and add four well-targeted test cases covering mixed-bot configs, empty allowed sets, stale configured-bot feedback, and no-bot-review scenarios.

_Note: Unable to post via the formal review endpoint due to a pre-existing pending review from a previous agent run._

**Comment by viamin** [comment]
Note: Unable to post formal PR review due to a pre-existing pending review. Posting review as issue comment.

**Review Summary**: The code changes are clean. The `allowed_review_bot_logins` extraction removes duplication between `check_review_bot_status` and `paid_agent_is_latest_blocker?`, and the nil-vs-empty-Set semantics are correct and well-documented in the helper comment. New tests cover meaningful edge cases (mixed-bot configs, empty allowed sets, address_all_bot_reviews interactions).

**Non-blocking note**: The PR description appears to be from a different PR — it references changes to `github_client.rb` and `prompts/build_for_pr.rb` (comment truncation, `max_prompt_comments`, etc.) that are not in this diff. Worth updating to match the actual changes.
